In [1]:
#manipulate dataframes in python
import pandas as pd
import numpy as np

#make API calls with python
import requests

#allows us to store results of API call cleanly
import json

import math
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

In [2]:
api_key = "a45a33bca80b5e1907adc4587be5c346b897dda7"

# Reason for choosing ACS1
The current main task is getting detailed population estimates for cities in the NYU City Health Dataset, which is the top 510 cities in the U.S. Therefore, ACS1 contains data for areas with populations of 65,000+. ACS1 will cover most cities in the NYU City Health Dataset. 

# Data check up

We want to get stable data variables throughout the years, so we can rely on the data that we pull out using API. <br>
The following will be checked first.<br>
1. Do 2010-2018 ACS have same cities -using FIPS CODE
2. Stable variable for different years -  Variables and/or variable names can be changed and it can give us different metrics (for example, one year is the whole number and one year is a percentage). We want to find the most stable variables for each category, so we can use the same metrics in the final data products. The variable test will be checked on each category. 

## 1. Do 2010-2018 ACS have same cities

We want to know ACS 1 year estimates of each year have the same cities. In the initial API pulls 2018, ACS1, which contains data for areas with populations of 65,000+, has 630 cities.  <br>

I will use two API data from Census 
1. ACS1 (Sex, Race, Ethnicity)
2. ACS1/subject (Age group)
<br>



ACS1 also has age, but ACS1/subject has age group. For example, ACS1 age is 1 or 5 years age groups, but the ACS1/subject age group is 15 to 44 and under 18. <br>So, we can minimize the calculation variables.

For checking number of cities in each dataset, I will check both dataset with total population and sex because these variables are present in both dataset and also variables in almost all cities, if the city has ACS dataset.

In [4]:
# For 

dic = {}
for year in range(2010,2019):
    dsource='acs'

    dname='acs1'
    sub = 'subject'
    cols='NAME,S0101_C01_001E'
    base_url = f'https://api.census.gov/data/{year}/{dsource}/{dname}/{sub}?get={cols}'
    
    
    # Looping over each states
    
    data_url= f'{base_url}&for=place:*&in=state:*&key={api_key}'
    response = requests.get(data_url)
    data=response.json()
    # Check how many place code in the each API
    print(year, len(data))
    dic[year] = data
    
    

2010 556
2011 563
2012 569
2013 583
2014 592
2015 597
2016 606
2017 615
2018 631


In [5]:
cols= ['NAME', 'Pop_Est', 'fips_state', 'fips_place']

In [6]:
# Set up each year's dataset

cities_2010 = pd.DataFrame(dic[2010][1:], columns=cols)
cities_2010["Year"]=2010


cities_2011 = pd.DataFrame(dic[2011][1:], columns=cols)
cities_2011["Year"]=2011


cities_2012 = pd.DataFrame(dic[2012][1:], columns=cols)
cities_2012["Year"]=2012

cities_2013 = pd.DataFrame(dic[2013][1:], columns=cols)
cities_2013["Year"]=2013


cities_2014 = pd.DataFrame(dic[2014][1:], columns=cols)
cities_2014["Year"]=2014


cities_2015 = pd.DataFrame(dic[2015][1:], columns=cols)
cities_2015["Year"]=2015

cities_2016 = pd.DataFrame(dic[2016][1:], columns=cols)
cities_2016["Year"]=2016

cities_2017 = pd.DataFrame(dic[2017][1:], columns=cols)
cities_2017["Year"]=2017

cities_2018 = pd.DataFrame(dic[2018][1:], columns=cols)
cities_2018["Year"]=2018

In [7]:
cities_2010

NAME  Pop_Est fips_state  \
0                             Birmingham city, Alabama   211670         01   
1                                 Dothan city, Alabama    66121         01   
2                                 Hoover city, Alabama    80376         01   
3                             Huntsville city, Alabama   181126         01   
4                                 Mobile city, Alabama   195249         01   
5                             Montgomery city, Alabama   206655         01   
6                             Tuscaloosa city, Alabama    90637         01   
7                       Anchorage municipality, Alaska   293227         02   
8                               Avondale city, Arizona    76446         04   
9                               Chandler city, Arizona   236775         04   
10                             Flagstaff city, Arizona    65976         04   
11                               Gilbert town, Arizona   209007         04   
12                              Glendale city, Arizona   227358         04   
13                              Goodyear city, Arizona    65445         04   
14                                  Mesa city, Arizona   440248         04   
15                                Peoria city, Arizona   154475         04   
16                               Phoenix city, Arizona  1449481         04   
17                         San Tan Valley CDP, Arizona    74977         04   
18                            Scottsdale city, Arizona   217977         04   
19                              Surprise city, Arizona   117826         04   
20                                 Tempe city, Arizona   162123         04   
21                                Tucson city, Arizona   521132         04   
22                                  Yuma city, Arizona    93409         04   
23                         Fayetteville city, Arkansas    73946         05   
24                           Fort Smith city, Arkansas    86338         05   
25                            Jonesboro city, Arkansas    67486         05   
26                          Little Rock city, Arkansas   193973         05   
27                           Springdale city, Arkansas    70751         05   
28                            Alameda city, California    73981         06   
29                           Alhambra city, California    83202         06   
30                            Anaheim city, California   337259         06   
31                            Antioch city, California   102745         06   
32                       Apple Valley town, California    69387         06   
33                        Arden-Arcade CDP, California    87272         06   
34                        Bakersfield city, California   348582         06   
35                       Baldwin Park city, California    75496         06   
36                         Bellflower city, California    76717         06   
37                           Berkeley city, California   112824         06   
38                         Buena Park city, California    80766         06   
39                            Burbank city, California   103472         06   
40                          Camarillo city, California    65406         06   
41                           Carlsbad city, California   105707         06   
42                             Carson city, California    91828         06   
43                              Chico city, California    86213         06   
44                              Chino city, California    78237         06   
45                        Chino Hills city, California    75070         06   
46                        Chula Vista city, California   244789         06   
47                     Citrus Heights city, California    83490         06   
48                             Clovis city, California    95915         06   
49                            Compton city, California    96558         06   
50                            Concord city, California   122508         

In [10]:
cities_in_2010 = cities_2010['NAME'].to_list()

In [11]:
# Check 2011

cities_2011[~cities_2011['NAME'].isin(cities_in_2010)]

NAME Pop_Est fips_state fips_place  Year
44         Glen Burnie CDP, Maryland   69026         24      32650  2011
313        Casas Adobes CDP, Arizona   67496         04      10670  2011
367  Florence-Graham CDP, California   68626         06      24477  2011
408       Palo Alto city, California   65419         06      55282  2011
414  Rancho Cordova city, California   65614         06      59444  2011
459    Walnut Creek city, California   65232         06      83346  2011
464       Yuba City city, California   65041         06      86972  2011

In [12]:
# Check 2012

cities_2012[~cities_2012['NAME'].isin(cities_in_2010)]

NAME Pop_Est fips_state fips_place  Year
1              Palo Alto city, California   66359         06      55282  2012
7         Rancho Cordova city, California   66999         06      59444  2012
37   South San Francisco city, California   65565         06      73262  2012
53          Walnut Creek city, California   65696         06      83346  2012
58             Yuba City city, California   65105         06      86972  2012
114               Pine Hills CDP, Florida   69219         12      56825  2012
216              Franklin city, Tennessee   66278         47      27740  2012
276                   Millcreek CDP, Utah   66939         49      50150  2012
306                Pasco city, Washington   64029         53      53545  2012
395             Glen Burnie CDP, Maryland   70285         24      32650  2012
471             Casas Adobes CDP, Arizona   71976         04      10670  2012
525       Florence-Graham CDP, California   63284         06      24477  2012
541        Jurupa Valley city, California   97419         06      37692  2012

In [13]:
# Check 2013

cities_2013[~cities_2013['NAME'].isin(cities_in_2010)]

NAME Pop_Est fips_state fips_place  Year
33           Bossier City city, Louisiana   66334         22      08920  2013
42                 Bethesda CDP, Maryland   66364         24      07125  2013
46            Gaithersburg city, Maryland   65689         24      31175  2013
48              Glen Burnie CDP, Maryland   66760         24      32650  2013
84                  Eagan city, Minnesota   65453         27      17288  2013
85            Maple Grove city, Minnesota   65406         27      40166  2013
91               Woodbury city, Minnesota   65659         27      71428  2013
118              Bayonne city, New Jersey   65024         34      03580  2013
158           Bismarck city, North Dakota   66063         38      07200  2013
204              Franklin city, Tennessee   68872         47      27740  2013
206          Johnson City city, Tennessee   65796         47      38320  2013
221             Kirkland city, Washington   84434         53      35940  2013
222                Pasco city, Washington   70345         53      53545  2013
295                  Victoria city, Texas   65097         48      75428  2013
327             Casas Adobes CDP, Arizona   69653         04      10670  2013
346      North Little Rock city, Arkansas   66086         05      50450  2013
362            Carmichael CDP, California   66747         06      11390  2013
398        Jurupa Valley city, California   98036         06      37692  2013
424            Palo Alto city, California   66638         06      55282  2013
430       Rancho Cordova city, California   67919         06      59444  2013
444         San Clemente city, California   65021         06      65084  2013
461  South San Francisco city, California   66174         06      73262  2013
462        South Whittier CDP, California   65365         06      73430  2013
478         Walnut Creek city, California   66890         06      83346  2013
483            Yuba City city, California   65410         06      86972  2013
526               Kissimmee city, Florida   65186         12      36950  2013
540               Pine Hills CDP, Florida   67295         12      56825  2013

In [14]:
# Check 2015

cities_2015[~cities_2015['NAME'].isin(cities_in_2010)]

NAME     Pop_Est fips_state fips_place  \
9               Casas Adobes CDP, Arizona       65265         04      10670   
28       North Little Rock city, Arkansas       66506         05      50450   
44             Carmichael CDP, California       63215         06      11390   
65        Florence-Graham CDP, California       66359         06      24477   
81         Jurupa Valley city, California      100314         06      37692   
82         Laguna Niguel city, California       65812         06      39248   
108            Palo Alto city, California       66848         06      55282   
114       Rancho Cordova city, California       71026         06      59444   
125               Brentwood CDP, New York       63792         36      08026   
149           Bismarck city, North Dakota       70240         38      07200   
195              Franklin city, Tennessee       72635         47      27740   
197          Johnson City city, Tennessee       65369         47      38320   
213                Cedar Park city, Texas       64837         48      13552   
215                    Conroe city, Texas       68592         48      16432   
241             New Braunfels city, Texas       69923         48      50820   
256                  Victoria city, Texas       67581         48      75428   
266               South Jordan city, Utah       66639         49      70850   
290             Kirkland city, Washington       87267         53      35940   
291           Marysville city, Washington       66773         53      43955   
292                Pasco city, Washington       68097         53      53545   
322         San Clemente city, California       65532         06      65084   
339  South San Francisco city, California       67271         06      73262   
355         Walnut Creek city, California       68906         06      83346   
360            Yuba City city, California       66944         06      86972   
364             Broomfield city, Colorado       65065         08      09280   
396            Delray Beach city, Florida       66261         12      17100   
403               Homestead city, Florida       66500         12      32275   
406               Kissimmee city, Florida       69152         12      36950   
420               Pine Hills CDP, Florida       76081         12      56825   
422                Poinciana CDP, Florida       73943         12      57900   
431             The Villages CDP, Florida  -999999999         12      71625   
441            Macon-Bibb County, Georgia      153459         13      49008   
480                       Ames city, Iowa       65052         19      01855   
491                  Shawnee city, Kansas       65058         20      64500   
497          Bossier City city, Louisiana       68097         22      08920   
509           Gaithersburg city, Maryland       67445         24      31175   
511             Glen Burnie CDP, Maryland       69859         24      32650   
512              Rockville city, Maryland       66998         24      67675   
548                 Eagan city, Minnesota       66288         27      17288   
549           Maple Grove city, Minnesota       68381         27      40166   
555              Woodbury city, Minnesota       67850         27      71428   
582              Bayonne city, New Jersey       66303         34      03580   

     Year  
9    2015  
28   2015  
44   2015  
65   2015  
81   2015  
82   2015  
108  2015  
114  2015  
125  2015  
149  2015  
195  2015  
197  2015  
213  2015  
215  2015  
241  2015  
256  2015  
266  2015  
290  2015  
291  2015  
292  2015  
322  2015  
339  2015  
355  2015  
360  2015  
364  2015  
396  2015  
403  2015  
406  2015  
420  2015  
422  2015  
431  2015  
441  2015  
480  2015  
491  2015  
497  2015  
509  2015  
511  2015  
512  2015  
548  2015  
549  2015  
555  2015  
582  2015

In [15]:
# Check 2018
cities_2018[~cities_2018['NAME'].isin(cities_in_2010)]

NAME     Pop_Est fips_state fips_place  \
12               Palm Harbor CDP, Florida       65994         12      54350   
13               Alpharetta city, Georgia       66263         13      01696   
24                     Conroe city, Texas       87656         48      16432   
25            Marysville city, Washington       69765         53      43955   
32             Delray Beach city, Florida       69356         12      17100   
37               Woodbury city, Minnesota       71299         27      71428   
38               Bayonne city, New Jersey       65082         34      03580   
45              New Braunfels city, Texas       85566         48      50820   
46                South Jordan city, Utah       74155         49      70850   
48                   Auburn city, Alabama       65737         01      03076   
53        Florence-Graham CDP, California       63621         06      24477   
60                        Ames city, Iowa       67155         19      01855   
62              Glen Burnie CDP, Maryland       65987         24      32650   
63              Lakeville city, Minnesota       65871         27      35180   
91              Casas Adobes CDP, Arizona       75694         04      10670   
115        South Whittier CDP, California       65577         06      73430   
126                Georgetown city, Texas       74176         48      29336   
131         Castro Valley CDP, California       63226         06      11964   
133                     Ankeny city, Iowa       65282         19      02305   
143            West Des Moines city, Iowa       67127         19      83910   
152                  Victoria city, Texas       67020         48      75428   
163               Homestead city, Florida       70469         12      32275   
165                  Shawnee city, Kansas       65844         20      64500   
181        Framingham city, Massachusetts       73123         25      24960   
186                       Lehi city, Utah       66029         49      44320   
190                 Conway city, Arkansas       66421         05      15190   
197         Walnut Creek city, California       69814         06      83346   
203               Kissimmee city, Florida       73596         12      36950   
204                Poinciana CDP, Florida       71331         12      57900   
212      North Little Rock city, Arkansas       66126         05      50450   
217             Broomfield city, Colorado       69267         08      09280   
239  South San Francisco city, California       67736         06      73262   
244           Maple Grove city, Minnesota       71812         27      40166   
249              Franklin city, Tennessee       80920         47      27740   
269        Jurupa Valley city, California      108400         06      37692   
273             The Hammocks CDP, Florida       64910         12      71569   
282            Sammamish city, Washington       65726         53      61115   
289            Yuba City city, California       66988         06      86972   
298               Brentwood CDP, New York       64855         36      08026   
316            South Fulton city, Georgia       97266         13      72122   
320          Bossier City city, Louisiana       68222         22      08920   
325           Bismarck city, North Dakota       71843         38      07200   
335               Madera city, California       65711         06      45022   
339                 Jupiter town, Florida       65544         12      35875   
341            Macon-Bibb County, Georgia      153095         13      49008   
352                 Lodi city, California       66995         06      42202   
370             Kirkland city, Washington       89557         53      35940   
377           Wellington village, Florida       65190         12      75812   
390                 Rogers city, Arkansas       67615         05      60410   
393        Lake Elsinore city, California       68187         06      39486   
394     

In [16]:
# Check 2018 with 2017

cities_2018[~cities_2018['NAME'].isin(cities_2017['NAME'])]

NAME Pop_Est fips_state fips_place  Year
12         Palm Harbor CDP, Florida   65994         12      54350  2018
48             Auburn city, Alabama   65737         01      03076  2018
63        Lakeville city, Minnesota   65871         27      35180  2018
115  South Whittier CDP, California   65577         06      73430  2018
133               Ankeny city, Iowa   65282         19      02305  2018
143      West Des Moines city, Iowa   67127         19      83910  2018
181  Framingham city, Massachusetts   73123         25      24960  2018
186                 Lehi city, Utah   66029         49      44320  2018
282      Sammamish city, Washington   65726         53      61115  2018
316      South Fulton city, Georgia   97266         13      72122  2018
339           Jupiter town, Florida   65544         12      35875  2018
377     Wellington village, Florida   65190         12      75812  2018
394        Rocklin city, California   67231         06      62364  2018
499             Rowlett city, Texas   66686         48      63572  2018
502          Blaine city, Minnesota   65215         27      06382  2018
511       Noblesville city, Indiana   66358         18      54180  2018
535      Wesley Chapel CDP, Florida   65249         12      75875  2018
581        Redmond city, Washington   67680         53      57535  2018

In [17]:
# Check ACS1 API


dic = {}
for year in range(2010,2019):
    dsource='acs'

    dname='acs1'
    sub = 'subject'
    cols='NAME,B01001_001E'
    base_url = f'https://api.census.gov/data/{year}/{dsource}/{dname}?get={cols}'
    
    
    # Looping over each states
    
    data_url= f'{base_url}&for=place:*&in=state:*&key={api_key}'
    response = requests.get(data_url)
    data=response.json()
    # Check how many place code in the each API
    print(year, len(data))
    dic[year] = data
    
    

2010 556
2011 563
2012 569
2013 583
2014 592
2015 597
2016 606
2017 615
2018 631




The following is the chart for the number of cities in ACS1 and ACS1/Subject, which are same for both API.


Year | Number of Cities
-- | --
2010 | 556
2011 | 563
2012 | 569
2013 | 583
2014 | 592
2015 | 597
2016 | 606
2017 | 615
2018 | 631

When I checked each year's cities that are added, most of the cities' populations are around 65,000. Therefore, it is added when the population estimates as 65,000. <bR>

I will check with other variables (SEX, RACE, Ethnicity) to find out how many cities that we want to use. Yet, my minimum will 556 because it is the number of cities from 2010.  <br>

I found out that one city with -999999, which is The Villages CDP, Florida	-999999999. I look for meaning. 

# Sex

For Sex, I want to find out right variable to use constantly throughout the year because in some year, it may give us proportion.<br>

For Sex, ACS has two main variables for SEX.

B01001_002E --> Estimate!!Total!!Male <Br>
B01001_026E --> Estimate!!Total!!Female

In [37]:

dic = {}
for year in range(2010,2019):
    dsource='acs'

    dname='acs1'
    sub = 'subject'
    cols='NAME,B01001_001E,B01001_002E,B01001_026E'
    base_url = f'https://api.census.gov/data/{year}/{dsource}/{dname}?get={cols}'
    
    
    # Looping over each states
    
    data_url= f'{base_url}&for=place:*&in=state:*&key={api_key}'
    response = requests.get(data_url)
    data=response.json()
    # Check how many place code in the each API
    print(year, len(data))
    dic[year] = data


2010 556
2011 563
2012 569
2013 583
2014 592
2015 597
2016 606
2017 615
2018 631


In [39]:
cols_sex= ['NAME', 'census_population','pop_male', 'pop_female','fips_state', 'fips_place']

# Set up each year's dataset

cities_2010 = pd.DataFrame(dic[2010][1:], columns=cols_sex)
cities_2010["Year"]=2010


cities_2011 = pd.DataFrame(dic[2011][1:], columns=cols_sex)
cities_2011["Year"]=2011


cities_2012 = pd.DataFrame(dic[2012][1:], columns=cols_sex)
cities_2012["Year"]=2012

cities_2013 = pd.DataFrame(dic[2013][1:], columns=cols_sex)
cities_2013["Year"]=2013


cities_2014 = pd.DataFrame(dic[2014][1:], columns=cols_sex)
cities_2014["Year"]=2014


cities_2015 = pd.DataFrame(dic[2015][1:], columns=cols_sex)
cities_2015["Year"]=2015

cities_2016 = pd.DataFrame(dic[2016][1:], columns=cols_sex)
cities_2016["Year"]=2016

cities_2017 = pd.DataFrame(dic[2017][1:], columns=cols_sex)
cities_2017["Year"]=2017

cities_2018 = pd.DataFrame(dic[2018][1:], columns=cols_sex)
cities_2018["Year"]=2018

In [40]:
cities_2010.head()

NAME census_population pop_male pop_female fips_state  \
0  Birmingham city, Alabama            211670    98760     112910         01   
1      Dothan city, Alabama             66121    31087      35034         01   
2      Hoover city, Alabama             80376    38694      41682         01   
3  Huntsville city, Alabama            181126    88132      92994         01   
4      Mobile city, Alabama            195249    91746     103503         01   

  fips_place  Year  
0      07000  2010  
1      21184  2010  
2      35896  2010  
3      37000  2010  
4      50000  2010

In [41]:
cities_2011.head()

NAME census_population pop_male pop_female  \
0               Reno city, Nevada            227509   117029     110480   
1             Sparks city, Nevada             91202    46436      44766   
2       Spring Valley CDP, Nevada            191336    94592      96744   
3       Sunrise Manor CDP, Nevada            181940    92688      89252   
4  Manchester city, New Hampshire            109835    56156      53679   

  fips_state fips_place  Year  
0         32      60600  2011  
1         32      68400  2011  
2         32      68585  2011  
3         32      71400  2011  
4         33      45140  2011

In [42]:
cities_2012.head()

NAME census_population pop_male pop_female fips_state  \
0  Birmingham city, Alabama            212835    97165     115670         01   
1      Dothan city, Alabama             68104    32389      35715         01   
2      Hoover city, Alabama             81627    39770      41857         01   
3  Huntsville city, Alabama            183076    91227      91849         01   
4      Mobile city, Alabama            194823    92531     102292         01   

  fips_place  Year  
0      07000  2012  
1      21184  2012  
2      35896  2012  
3      37000  2012  
4      50000  2012

In [43]:
cities_2013.head()

NAME census_population pop_male pop_female  \
0  Chattanooga city, Tennessee            173375    82956      90419   
1     Birmingham city, Alabama            211933   102318     109615   
2         Dothan city, Alabama             67296    32208      35088   
3         Hoover city, Alabama             84128    38013      46115   
4     Huntsville city, Alabama            186416    91771      94645   

  fips_state fips_place  Year  
0         47      14000  2013  
1         01      07000  2013  
2         01      21184  2013  
3         01      35896  2013  
4         01      37000  2013

In [44]:
cities_2014.head()

NAME census_population pop_male pop_female fips_state  \
0  Birmingham city, Alabama            212653    98257     114396         01   
1      Dothan city, Alabama             69400    32542      36858         01   
2      Hoover city, Alabama             84352    40334      44018         01   
3  Huntsville city, Alabama            187592    90041      97551         01   
4      Mobile city, Alabama            194670    92585     102085         01   

  fips_place  Year  
0      07000  2014  
1      21184  2014  
2      35896  2014  
3      37000  2014  
4      50000  2014

In [45]:
cities_2015.head()

NAME census_population pop_male pop_female fips_state  \
0  Birmingham city, Alabama            214911   102122     112789         01   
1      Dothan city, Alabama             67536    32172      35364         01   
2      Hoover city, Alabama             84839    38040      46799         01   
3  Huntsville city, Alabama            189114    91764      97350         01   
4      Mobile city, Alabama            194305    91275     103030         01   

  fips_place  Year  
0      07000  2015  
1      21184  2015  
2      35896  2015  
3      37000  2015  
4      50000  2015

In [46]:
cities_2016.head()

NAME census_population pop_male pop_female fips_state  \
0  Birmingham city, Alabama            213434   100602     112832         01   
1      Dothan city, Alabama             67714    32089      35625         01   
2      Hoover city, Alabama             84943    40646      44297         01   
3  Huntsville city, Alabama            196225    96482      99743         01   
4      Mobile city, Alabama            192895    92787     100108         01   

  fips_place  Year  
0      07000  2016  
1      21184  2016  
2      35896  2016  
3      37000  2016  
4      50000  2016

In [47]:
cities_2017.head()

NAME census_population pop_male pop_female  \
0           Urban Honolulu CDP, Hawaii            350388   172554     177834   
1               Boise City city, Idaho            226567   115099     111468   
2                 Meridian city, Idaho             99938    49540      50398   
3                    Nampa city, Idaho             93585    48680      44905   
4  Arlington Heights village, Illinois             75097    35849      39248   

  fips_state fips_place  Year  
0         15      71550  2017  
1         16      08830  2017  
2         16      52120  2017  
3         16      56260  2017  
4         17      02154  2017

In [48]:
cities_2018.head()

NAME census_population pop_male pop_female  \
0  Lawrence city, Massachusetts             80370    38578      41792   
1      Clifton city, New Jersey             85272    41941      43331   
2              Parma city, Ohio             78746    39251      39495   
3         Lawrence city, Kansas             97293    48090      49203   
4          Cleveland city, Ohio            383781   188481     195300   

  fips_state fips_place  Year  
0         25      34550  2018  
1         34      13690  2018  
2         39      61000  2018  
3         20      38900  2018  
4         39      16000  2018

In [49]:

# Combine all AGE group
frames = [cities_2010, cities_2011, cities_2012, cities_2013,cities_2014,cities_2015,cities_2016, cities_2017, cities_2018]

In [50]:
age_result = pd.concat(frames)

In [51]:
age_result.head()

NAME census_population pop_male pop_female fips_state  \
0  Birmingham city, Alabama            211670    98760     112910         01   
1      Dothan city, Alabama             66121    31087      35034         01   
2      Hoover city, Alabama             80376    38694      41682         01   
3  Huntsville city, Alabama            181126    88132      92994         01   
4      Mobile city, Alabama            195249    91746     103503         01   

  fips_place  Year  
0      07000  2010  
1      21184  2010  
2      35896  2010  
3      37000  2010  
4      50000  2010

In [56]:
age_result['fips_state_place']=age_result['fips_state']+age_result['fips_place']

In [57]:
age_result.tail()

NAME census_population pop_male pop_female  \
625         Menifee city, California             92602    46509      46093   
626       Riverside city, California            330080   160110     169970   
627  Springfield city, Massachusetts            155029    75821      79208   
628          Dale City CDP, Virginia             68805    35840      32965   
629      South Gate city, California             94439    47917      46522   

    fips_state fips_place  Year fips_state_place  
625         06      46842  2018          0646842  
626         06      62000  2018          0662000  
627         25      67000  2018          2567000  
628         51      21088  2018          5121088  
629         06      73080  2018          0673080

In [54]:
# Creating NYU_500_Largest to find out whether NYU dataset has this city or not.

# Pull out NYU FIPS 
# Preselect column type as object
coltype = {'fips_state':object,
          'fips_place':object,
           'fips_state_place':object}
nyu = pd.read_csv(r"/Users/jasonlim/Desktop/EFGS/Census-master/CitiesDataset/City_Health_data/nyu_fips_code.csv",\
                  dtype= coltype)


In [55]:
nyu.head()

index state_abb fips_state fips_place fips_state_place   city_name
0      0        HI         15      00003          0015003    Honolulu
1      1        AL         01      07000          0107000  Birmingham
2      2        AL         01      35896          0135896      Hoover
3      3        AL         01      37000          0137000  Huntsville
4      4        AL         01      50000          0150000      Mobile

In [61]:
nyu_cities = list(nyu['fips_state_place'])

In [63]:
age_result['NYU_500_Largest'] = age_result['fips_state_place'].apply(lambda x:1 if x in nyu_cities else 0)

In [64]:
age_result.head()

NAME census_population pop_male pop_female fips_state  \
0  Birmingham city, Alabama            211670    98760     112910         01   
1      Dothan city, Alabama             66121    31087      35034         01   
2      Hoover city, Alabama             80376    38694      41682         01   
3  Huntsville city, Alabama            181126    88132      92994         01   
4      Mobile city, Alabama            195249    91746     103503         01   

  fips_place  Year fips_state_place  NYU_500_Largest  
0      07000  2010          0107000                1  
1      21184  2010          0121184                0  
2      35896  2010          0135896                1  
3      37000  2010          0137000                1  
4      50000  2010          0150000                1

In [65]:
age_result.to_csv("wave2_first.csv", index=False)

## Checking How many NYU cities was covered in each year

Overall average is 0.84, which means that it covered 84% of NYU cities.

In [66]:
age_result.groupby('Year').sum()

NYU_500_Largest
Year                 
2010              496
2011              496
2012              496
2013              496
2014              495
2015              495
2016              495
2017              495
2018              495

As shown above, from 509 cities of NYU dataset, 495/ 496 cities are covered by ACS1. Most of cities are covered by ACS1.

# Jason's work until 6/4


Below codes will ref. for future work.


# On Monday, 6/8,

To-do
1. Check Race, Ethnicity.
2. Age group

# Data Sources

I will use Census American Community Survey for the following categories:


1. Age: pop_18+
2. Age: pop_65+
3. pop_18+pe
4. pop_65+pe
5. Sex: pop_male
6. Sex:pop_female
7. pop_male_pe
8. pop_female_pe

10. Race: pop_white
11. pop_white_pe
12. Race: pop__black
13. pop__black_pe
14. Race: pop__asian
15. pop__asian_pe
16. Race: pop__AIAN
17. pop__AIAN_pe
18. Race: pop__NHOPI
19. pop__NHOPI_pe
20. Ethnicity: pop__nonhispanic
21. Ethnicity: pop__hispanic

For pe (percentage), I will get the total ACS population estimate first. Then,in each level categories (AGE, SEX, RACE, Ethnicity) and get percentage of each fields. 



In [3]:
# Sample 

year='2018'
dsource='acs'

dname='acs1'
sub = 'subject'
cols='NAME,S0101_C01_001E,S0101_C01_026E,S0101_C01_030E'

state='36'
place='51000'

base_url = f'https://api.census.gov/data/{year}/{dsource}/{dname}/{sub}'

data_url= f'{base_url}?get={cols}&for=place:{place}&in=state:{state}&key={api_key}'
print(data_url)
response = requests.get(data_url)

print(response.text)


https://api.census.gov/data/2018/acs/acs1/subject?get=NAME,S0101_C01_001E,S0101_C01_026E,S0101_C01_030E&for=place:51000&in=state:36&key=a45a33bca80b5e1907adc4587be5c346b897dda7
[["NAME","S0101_C01_001E","S0101_C01_026E","S0101_C01_030E","state","place"],
["New York city, New York","8398748","6659492","1245480","36","51000"]]


# AGE

Age we want is 18 years or over and 65 years or over. 
In Census API data, ACS1/subject has <br>
S0101_C01_026E	Estimate!!Total!!Total population!!SELECTED AGE CATEGORIES!!18 years and over <br>
S0101_C01_030E	Estimate!!Total!!Total population!!SELECTED AGE CATEGORIES!!65 years and over <br>

Note: ACS1 has age but it does not give me 18+ or 65+.

In [4]:
# 2018 AGE DAta

year='2018'
dsource='acs'

dname='acs1'
sub = 'subject'

base_url = f'https://api.census.gov/data/{year}/{dsource}/{dname}/{sub}?get={cols}'

cols='NAME,S0101_C01_001E,S0101_C01_026E,S0101_C01_030E'
# Looping over each states

data_url= f'{base_url}&for=place:*&in=state:*&key={api_key}'
response = requests.get(data_url)
data=response.json()



In [5]:
age_2018 = pd.DataFrame(data[1:], columns=['city_name', 'Total_pop_ACS', 'pop_18+', 'pop_65+', 'fips_state', 'fips_place'])


In [6]:
age_2018.head()

city_name Total_pop_ACS pop_18+ pop_65+ fips_state  \
0                Yuma city, Arizona         97917   72990   15465         04   
1          Concord city, California        129681  107596   20253         06   
2  East Los Angeles CDP, California        117961   89038   13381         06   
3         El Cajon city, California        103229   77139   12503         06   
4             Napa city, California         79261   61126   14605         06   

  fips_place  
0      85540  
1      16000  
2      20802  
3      21712  
4      50258

In [7]:
# DATA Pre-processing

age_2018['Total_pop_ACS'] = age_2018['Total_pop_ACS'].astype('int')
age_2018['pop_18+'] = age_2018['pop_18+'].astype('int')
age_2018['pop_65+'] = age_2018['pop_65+'].astype('int')

# Combining STATE and PLACE FIPS
age_2018['fips_state_place'] = age_2018['fips_state']+age_2018['fips_place']

# Calculate percentage

age_2018['pop_18+_pe']=100*round(age_2018['pop_18+']/age_2018['Total_pop_ACS'],4)

age_2018['pop_65+_pe']=100*round(age_2018['pop_65+']/age_2018['Total_pop_ACS'],4)

In [8]:
age_2018.head()

city_name  Total_pop_ACS  pop_18+  pop_65+  \
0                Yuma city, Arizona          97917    72990    15465   
1          Concord city, California         129681   107596    20253   
2  East Los Angeles CDP, California         117961    89038    13381   
3         El Cajon city, California         103229    77139    12503   
4             Napa city, California          79261    61126    14605   

  fips_state fips_place fips_state_place  pop_18+_pe  pop_65+_pe  
0         04      85540          0485540       74.54       15.79  
1         06      16000          0616000       82.97       15.62  
2         06      20802          0620802       75.48       11.34  
3         06      21712          0621712       74.73       12.11  
4         06      50258          0650258       77.12       18.43

In [9]:
age_2018.to_csv("age_2018.csv", index=False)

In [10]:
age_2018.columns

Index(['city_name', 'Total_pop_ACS', 'pop_18+', 'pop_65+', 'fips_state',
       'fips_place', 'fips_state_place', 'pop_18+_pe', 'pop_65+_pe'],
      dtype='object')

In [11]:
# Remove city name from age dataset because it can cause error by small differences such as space between letter.

age_2018= age_2018[['fips_state_place', 'Total_pop_ACS', 'pop_18+', 'pop_65+','pop_18+_pe', 'pop_65+_pe' ]]

In [12]:
# for getting place ID, I will pull up wave1 data.
coltype1 = {'fips_state':object,
           'fips_place':object,
           'fips_state_place': object}
wave1= pd.read_excel("wave1.xlsx",dtype=coltype1)

In [13]:
wave1.head()

fips_state fips_place fips_state_place        city_name  st_name  \
0         01      00124          0100124   Abbeville city  Alabama   
1         01      00460          0100460  Adamsville city  Alabama   
2         01      00484          0100484     Addison town  Alabama   
3         01      00676          0100676       Akron town  Alabama   
4         01      00820          0100820   Alabaster city  Alabama   

   NYU_500_Largest  year  census_population  
0                0  2018               2563  
1                0  2018               4325  
2                0  2018                724  
3                0  2018                330  
4                0  2018              33340

In [14]:
# combine with first dataset

wave_age = pd.merge(wave1, age_2018, how='left',on='fips_state_place')

In [15]:
wave_age.head()

fips_state fips_place fips_state_place        city_name  st_name  \
0         01      00124          0100124   Abbeville city  Alabama   
1         01      00460          0100460  Adamsville city  Alabama   
2         01      00484          0100484     Addison town  Alabama   
3         01      00676          0100676       Akron town  Alabama   
4         01      00820          0100820   Alabaster city  Alabama   

   NYU_500_Largest  year  census_population  Total_pop_ACS  pop_18+  pop_65+  \
0                0  2018               2563            NaN      NaN      NaN   
1                0  2018               4325            NaN      NaN      NaN   
2                0  2018                724            NaN      NaN      NaN   
3                0  2018                330            NaN      NaN      NaN   
4                0  2018              33340            NaN      NaN      NaN   

   pop_18+_pe  pop_65+_pe  
0         NaN         NaN  
1         NaN         NaN  
2         NaN         NaN  
3         NaN         NaN  
4         NaN         NaN

In [16]:
# Checking Yuma City for POP ACS because it can be different
wave_age[wave_age['fips_state_place']=='0485540']

fips_state fips_place fips_state_place  city_name  st_name  \
700         04      85540          0485540  Yuma city  Arizona   

     NYU_500_Largest  year  census_population  Total_pop_ACS  pop_18+  \
700                1  2018              97908        97917.0  72990.0   

     pop_65+  pop_18+_pe  pop_65+_pe  
700  15465.0       74.54       15.79

In [17]:
wave_age['fips_state_place'].count()

19495

In [18]:

wave_age['pop_18+'].count()

582

In [19]:
# fill missing value with N/A code, -998

wave_age.fillna("-998",inplace=True)

# Gender, RACE, Ethnicity

For These categories, I can get it from ACS1 API.

https://api.census.gov/data/2018/acs/acs1

In [20]:
# 2018 AGE DAta

year='2018'
dsource='acs'

dname='acs1'

cols='NAME,B01001_001E,B01001_002E,B01001_026E,B02001_002E,B02001_003E,B02001_005E,\
B02001_004E,B02001_006E,B01001I_001E'
base_url = f'https://api.census.gov/data/{year}/{dsource}/{dname}/?get={cols}'


# Looping over each states

data_url= f'{base_url}&for=place:*&in=state:*&key={api_key}'
print(data_url)
response = requests.get(data_url)
data=response.json()


https://api.census.gov/data/2018/acs/acs1/?get=NAME,B01001_001E,B01001_002E,B01001_026E,B02001_002E,B02001_003E,B02001_005E,B02001_004E,B02001_006E,B01001I_001E&for=place:*&in=state:*&key=a45a33bca80b5e1907adc4587be5c346b897dda7


In [21]:
col=['city_name', 'Total_pop_GRE', 'pop_male', 'pop_female', \
     'pop_white','pop_black','pop_asian', 'pop_AIAN', 'pop_NHOPI','pop_hispanic', 'state', 'place']

In [22]:
GRE_2018 = pd.DataFrame(data[1:], columns=col)


In [25]:
GRE_2018.head()

city_name Total_pop_GRE pop_male pop_female pop_white  \
0      Auburn city, Alabama         65737    31482      34255     48691   
1  Birmingham city, Alabama        209294    96634     112660     56509   
2      Dothan city, Alabama         67814    32283      35531     43412   
3      Hoover city, Alabama         85115    39998      45117     60576   
4  Huntsville city, Alabama        199808    95202     104606    114924   

  pop_black pop_asian pop_AIAN pop_NHOPI pop_hispanic state  place  
0      9784      5671        0         0         None    01  03076  
1    141377      2946      361       206        10074    01  07000  
2     22262       603       53         0         None    01  21184  
3     14801      4687      198        43         None    01  35896  
4     62225      6221     1300         0        13397    01  37000

In [23]:
# CHECKING YUMA CITY FOR pop total. ALL ACS have same total pop for each city.
# So, in the end, I will use just one from pop total
GRE_2018[GRE_2018['city_name']=='Yuma city, Arizona']

city_name Total_pop_GRE pop_male pop_female pop_white pop_black  \
25  Yuma city, Arizona         97917    49808      48109     81988      3018   

   pop_asian pop_AIAN pop_NHOPI pop_hispanic state  place  
25      3077      850         0        58553    04  85540

In [24]:
# DATA Pre-processing
# None mean not available, because if there is none, it shows 0. 
GRE_2018.fillna(value=pd.np.nan, inplace=True)


# change to float NOTE: NOT to integer because we have nan value

GRE_2018['Total_pop_GRE']= GRE_2018['Total_pop_GRE'].astype('float')
GRE_2018['pop_male'] = GRE_2018['pop_male'].astype('float')
GRE_2018['pop_female'] = GRE_2018['pop_female'].astype('float')
GRE_2018['pop_white'] = GRE_2018['pop_white'].astype('float')
GRE_2018['pop_black'] = GRE_2018['pop_black'].astype('float')
GRE_2018['pop_asian'] = GRE_2018['pop_asian'].astype('float')
GRE_2018['pop_AIAN'] = GRE_2018['pop_AIAN'].astype('float')
GRE_2018['pop_NHOPI'] = GRE_2018['pop_NHOPI'].astype('float')
GRE_2018['pop_hispanic']= GRE_2018['pop_hispanic'].astype('float')


# Combining STATE and PLACE FIPS
GRE_2018['fips_state_place'] = GRE_2018['state']+GRE_2018['place']

# Calculate Non Hispanic
GRE_2018['pop_nonhispanic'] = GRE_2018['Total_pop_GRE'] - GRE_2018['pop_hispanic']


# Calculate percentage
GRE_2018['pop_male_pe']        = 100*round(GRE_2018['pop_male']/GRE_2018['Total_pop_GRE'], 4)
GRE_2018['pop_female_pe']      = 100*round(GRE_2018['pop_female']/GRE_2018['Total_pop_GRE'], 4)
GRE_2018['pop_white_pe']       = 100*round(GRE_2018['pop_white']/GRE_2018['Total_pop_GRE'], 4)
GRE_2018['pop_black_pe']       = 100*round(GRE_2018['pop_black']/GRE_2018['Total_pop_GRE'], 4)  
GRE_2018['pop_asian_pe']       = 100*round(GRE_2018['pop_asian']/GRE_2018['Total_pop_GRE'], 4)  
GRE_2018['pop_AIAN_pe']        = 100*round(GRE_2018['pop_AIAN']/GRE_2018['Total_pop_GRE'], 4)  
GRE_2018['pop_NHOPI_pe']       = 100*round(GRE_2018['pop_NHOPI']/GRE_2018['Total_pop_GRE'], 4)   
GRE_2018['pop_hispanic_pe']    = 100*round(GRE_2018['pop_hispanic']/GRE_2018['Total_pop_GRE'], 4)  
GRE_2018['pop_nonhispanic_pe'] = 100*round(GRE_2018['pop_nonhispanic']/GRE_2018['Total_pop_GRE'], 4)

/Users/jasonlim/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:3: FutureWarning: The pandas.np module is deprecated and will be removed from pandas in a future version. Import numpy directly instead
  This is separate from the ipykernel package so we can avoid doing imports until


In [25]:
GRE_2018.head()

city_name  Total_pop_GRE  pop_male  pop_female  pop_white  \
0      Auburn city, Alabama        65737.0   31482.0     34255.0    48691.0   
1  Birmingham city, Alabama       209294.0   96634.0    112660.0    56509.0   
2      Dothan city, Alabama        67814.0   32283.0     35531.0    43412.0   
3      Hoover city, Alabama        85115.0   39998.0     45117.0    60576.0   
4  Huntsville city, Alabama       199808.0   95202.0    104606.0   114924.0   

   pop_black  pop_asian  pop_AIAN  pop_NHOPI  pop_hispanic state  place  \
0     9784.0     5671.0       0.0        0.0           NaN    01  03076   
1   141377.0     2946.0     361.0      206.0       10074.0    01  07000   
2    22262.0      603.0      53.0        0.0           NaN    01  21184   
3    14801.0     4687.0     198.0       43.0           NaN    01  35896   
4    62225.0     6221.0    1300.0        0.0       13397.0    01  37000   

  fips_state_place  pop_nonhispanic  pop_male_pe  pop_female_pe  pop_white_pe  \
0          0103076              NaN        47.89          52.11         74.07   
1          0107000         199220.0        46.17          53.83         27.00   
2          0121184              NaN        47.61          52.39         64.02   
3          0135896              NaN        46.99          53.01         71.17   
4          0137000         186411.0        47.65          52.35         57.52   

   pop_black_pe  pop_asian_pe  pop_AIAN_pe  pop_NHOPI_pe  pop_hispanic_pe  \
0         14.88          8.63         0.00          0.00              NaN   
1         67.55          1.41         0.17          0.10             4.81   
2         32.83          0.89         0.08          0.00              NaN   
3         17.39          5.51         0.23          0.05              NaN   
4         31.14          3.11         0.65          0.00             6.70   

   pop_nonhispanic_pe  
0                 NaN  
1               95.19  
2                 NaN  
3                 NaN  
4               93.30

In [26]:
GRE_2018.columns

Index(['city_name', 'Total_pop_GRE', 'pop_male', 'pop_female', 'pop_white',
       'pop_black', 'pop_asian', 'pop_AIAN', 'pop_NHOPI', 'pop_hispanic',
       'state', 'place', 'fips_state_place', 'pop_nonhispanic', 'pop_male_pe',
       'pop_female_pe', 'pop_white_pe', 'pop_black_pe', 'pop_asian_pe',
       'pop_AIAN_pe', 'pop_NHOPI_pe', 'pop_hispanic_pe', 'pop_nonhispanic_pe'],
      dtype='object')

In [27]:
# RE-Organize for file export

GRE_2018=GRE_2018[['city_name', 'Total_pop_GRE', 'pop_male', 'pop_female', 'pop_male_pe','pop_female_pe',\
                   'pop_white','pop_black', 'pop_asian', 'pop_AIAN',  'pop_NHOPI',\
                   'pop_white_pe', 'pop_black_pe', 'pop_asian_pe','pop_AIAN_pe','pop_NHOPI_pe',\
                   'pop_hispanic','pop_nonhispanic', 'pop_hispanic_pe','pop_nonhispanic_pe',\
                   'state', 'place', 'fips_state_place']]
GRE_2018.to_csv("Gender_RACE_Ethnicity_2018.csv", index=False)

In [28]:
# Check whether total_pop_GRE has same number of ACS pop est
GRE_2018[GRE_2018['city_name']=='Concord city, California']

city_name  Total_pop_GRE  pop_male  pop_female  \
58  Concord city, California       129681.0   62538.0     67143.0   

    pop_male_pe  pop_female_pe  pop_white  pop_black  pop_asian  pop_AIAN  \
58        48.22          51.78    80258.0     4753.0    14189.0    1383.0   

    pop_NHOPI  pop_white_pe  pop_black_pe  pop_asian_pe  pop_AIAN_pe  \
58      156.0         61.89          3.67         10.94         1.07   

    pop_NHOPI_pe  pop_hispanic  pop_nonhispanic  pop_hispanic_pe  \
58          0.12       39740.0          89941.0            30.64   

    pop_nonhispanic_pe state  place fips_state_place  
58               69.36    06  16000          0616000

In [29]:
# Remove city name and total_pop_GRE from GRE dataset because it can cause error by small differences such as space between letter.


GRE_2018_a=GRE_2018[['fips_state_place', 'pop_male', 'pop_female', 'pop_male_pe','pop_female_pe',\
                   'pop_white','pop_black', 'pop_asian', 'pop_AIAN',  'pop_NHOPI',\
                   'pop_white_pe', 'pop_black_pe', 'pop_asian_pe','pop_AIAN_pe','pop_NHOPI_pe',\
                   'pop_hispanic','pop_nonhispanic', 'pop_hispanic_pe','pop_nonhispanic_pe']]

In [30]:
wave_2_2018 = pd.merge(wave_age , GRE_2018_a, how='left',on='fips_state_place')

In [31]:
wave_2_2018[wave_2_2018['fips_state_place']=='0485540']

fips_state fips_place fips_state_place  city_name  st_name  \
700         04      85540          0485540  Yuma city  Arizona   

     NYU_500_Largest  year  census_population Total_pop_ACS pop_18+ pop_65+  \
700                1  2018              97908         97917   72990   15465   

    pop_18+_pe pop_65+_pe  pop_male  pop_female  pop_male_pe  pop_female_pe  \
700      74.54      15.79   49808.0     48109.0        50.87          49.13   

     pop_white  pop_black  pop_asian  pop_AIAN  pop_NHOPI  pop_white_pe  \
700    81988.0     3018.0     3077.0     850.0        0.0         83.73   

     pop_black_pe  pop_asian_pe  pop_AIAN_pe  pop_NHOPI_pe  pop_hispanic  \
700          3.08          3.14         0.87           0.0       58553.0   

     pop_nonhispanic  pop_hispanic_pe  pop_nonhispanic_pe  
700          39364.0             59.8                40.2

In [32]:
wave_2_2018.fillna("-998",inplace=True)

In [ ]:
wave_2_2018.to_csv("place_pop_2018.csv", index=False)
wave_2_2018.to_excel("place_pop_2018.xlsx", index=False)